Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -918,6 +966,146 @@ lemma nonsingular [Nontrivial R] {x y : R} (h : E.equation x y) : E.nonsingular
   E.nonsingular_of_Δ_ne_zero h <| E.coe_Δ' ▸ E.Δ'.ne_zero
 #align elliptic_curve.nonsingular EllipticCurve.nonsingular
 
+section ModelsWithJInvariant_0_1728
+
+variable (R : Type u) [CommRing R]
+
+/-- The elliptic curve $Y^2 + Y = X^3$ when $3$ is invertible. -/
+def ofJInvariant0 [Invertible (3 : R)] : EllipticCurve R :=
+  ⟨
+    WeierstrassCurve.ofJInvariant0 R,
+    ⟨-27, -(⅟3) ^ 3,
+      by
+        rw [calc (27 : R) = (3 : R) ^ 3 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 3 (mul_invOf_self (3 : R))
+      , by
+        rw [calc (27 : R) = (3 : R) ^ 3 := by norm_num]
+        simp only [mul_neg, neg_mul, neg_neg]
+        exact pow_mul_pow_eq_one 3 (invOf_mul_self (3 : R))
+    ⟩,
+    by rw [WeierstrassCurve.ofJInvariant0_Δ R]
+  ⟩
```

Comment:
The weird convention in mathlib is to not have brackets alone in a line, and commas at the end of lines rather than at the beginning.
```suggestion
/-- The elliptic curve $Y^2 + Y = X^3$ when $3$ is invertible. -/
def ofJInvariant0 [Invertible (3 : R)] : EllipticCurve R :=
  ⟨WeierstrassCurve.ofJInvariant0 R,
    ⟨-27, -⅟3 ^ 3,
      by rw [neg_mul_neg, show (27 : R) = 3 ^ 3 by norm_num1,
        pow_mul_pow_eq_one 3 <| mul_invOf_self 3],
      by rw [neg_mul_neg, show (27 : R) = 3 ^ 3 by norm_num1,
        pow_mul_pow_eq_one 3 <| invOf_mul_self 3]⟩,
    by rw [WeierstrassCurve.ofJInvariant0_Δ R]⟩
```

Reply:
Cool. Are there more compact ways to show -27 is invertible?

Reply:
This works:
```
def ofJ0 [Invertible (3 : R)] : EllipticCurve R :=
  have := invertibleNeg (3 ^ 3 : R)
  ⟨WeierstrassCurve.ofJ0 R, unitOfInvertible (-3 ^ 3 : R),
    by rw [unitOfInvertible_val, WeierstrassCurve.ofJ0_Δ R]; norm_num1⟩
```

Reply:
> The lemmas `ofJ0_j`, `ofJ1728_j`, and `ofJ'_j` might need `unitOfInvertible_val`, and with this can mostly be `norm_num1`'ed now.

I cannot make `ofJ1728_j` and `ofJ'_j` work, `rw` refuses to rewrite the result obtained from `unitOfInvertible_val`, although they look the same on the info panel.